provisioning/image_customization.go

```diff
@@ -0,0 +1,260 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provisioning
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	metal3iov1alpha1 "github.com/openshift/cluster-baremetal-operator/api/v1alpha1"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+)
+
+const (
+	ironicBaseUrl                    = "IRONIC_BASE_URL"
+	ironicAgentImage                 = "IRONIC_AGENT_IMAGE"
+	imageCustomizationDeploymentName = "metal3-image-customization"
+	imageCustomizationVolume         = "metal3-image-customization-volume"
+	imageCustomizationPort           = 8084
+	containerRegistriesConfPath      = "/etc/containers/registries.conf"
+	containerRegistriesEnvVar        = "REGISTRIES_CONF_PATH"
+	deployISOEnvVar                  = "DEPLOY_ISO"
+	deployISOFile                    = "/shared/html/images/ironic-python-agent.iso"
+	deployInitrdEnvVar               = "DEPLOY_INITRD"
+	deployInitrdFile                 = "/shared/html/images/ironic-python-agent.initramfs"
+)
+
+var (
+	imageRegistriesVolumeMount = corev1.VolumeMount{
+		Name:      imageCustomizationVolume,
+		MountPath: containerRegistriesConfPath,
+	}
+)
+
+func imageRegistriesVolume() corev1.Volume {
+	// TODO: Should this be corev1.HostPathFile?
+	volType := corev1.HostPathFileOrCreate
+
+	return corev1.Volume{
+		Name: imageCustomizationVolume,
+		VolumeSource: corev1.VolumeSource{
+			HostPath: &corev1.HostPathVolumeSource{
+				Path: containerRegistriesConfPath,
+				Type: &volType,
+			},
+		},
+	}
+}
+
+func hostForIP(ipAddr string) string {
```

Comment:
Given that this is now only being called once, and we probably want some logic to skip adding the "http://" if it's empty, maybe we should just inline this logic in `createImageCustomizationContainer`.